pkg/agent/controller/networkpolicy/fqdn_test.go

```diff
@@ -0,0 +1,286 @@
+			}
+			if tt.existingSelectorToRuleIDs != nil {
+				f.selectorItemToRuleIDs = tt.existingSelectorToRuleIDs
+				f.fqdnToSelectorItem = tt.existingFQDNToSelectorItem
+			}
+			if tt.existingDNSCache != nil {
+				f.dnsEntryCache = tt.existingDNSCache
+			}
+			if err := f.addFQDNRule(tt.ruleID, tt.fqdns, tt.podAddrs); err != nil {
+				t.Fatalf("Error occurred in addFQDNRule %v", err)
+			}
+			assert.Equal(t, tt.finalSelectorToRuleIDs, f.selectorItemToRuleIDs)
+			assert.Equal(t, tt.finalFQDNToSelectorItem, f.fqdnToSelectorItem)
+		})
+	}
+}
+
+type fqdnRuleAddArgs struct {
+	ruleID         string
+	fqdns          []string
+	podOFAddresses sets.Int32
+}
+
+func TestDeleteFQDNRule(t *testing.T) {
+	selectorItem1 := fqdnSelectorItem{
+		matchName: "test.antrea.io",
+	}
+	selectorItem2 := fqdnSelectorItem{
+		matchRegex: "^.*antrea[.]io$",
+	}
+	selectorItem3 := fqdnSelectorItem{
+		matchName: "maps.google.com",
+	}
+	tests := []struct {
+		name                    string
+		previouslyAddedRules    []fqdnRuleAddArgs
+		existingDNSCache        map[string]dnsMeta
+		ruleID                  string
+		fqdns                   []string
+		finalSelectorToRuleIDs  map[fqdnSelectorItem]sets.String
+		finalFQDNToSelectorItem map[string]map[fqdnSelectorItem]struct{}
+		addressRemoved          bool
+	}{
+		{
+			"test-1",
+			[]fqdnRuleAddArgs{
+				{
+					"mockRule1",
+					[]string{"test.antrea.io"},
+					sets.NewInt32(1),
+				},
+			},
+			map[string]dnsMeta{
+				"test.antrea.io": {},
+			},
+			"mockRule1",
+			[]string{"test.antrea.io"},
+			map[fqdnSelectorItem]sets.String{},
+			map[string]map[fqdnSelectorItem]struct{}{},
+			true,
+		},
+		{
+			"test-2",
+			[]fqdnRuleAddArgs{
+				{
+					"mockRule1",
+					[]string{"test.antrea.io"},
+					sets.NewInt32(1),
+				},
+				{
+					"mockRule2",
+					[]string{"test.antrea.io"},
+					sets.NewInt32(2),
+				},
+			},
+			map[string]dnsMeta{
+				"test.antrea.io": {},
+			},
+			"mockRule1",
+			[]string{"test.antrea.io"},
+			map[fqdnSelectorItem]sets.String{
+				selectorItem1: sets.NewString("mockRule2"),
+			},
+			map[string]map[fqdnSelectorItem]struct{}{
+				"test.antrea.io": {
+					selectorItem1: struct{}{},
+				},
+			},
+			true,
+		},
+		{
+			"test-3",
+			[]fqdnRuleAddArgs{
+				{
+					"mockRule1",
+					[]string{"test.antrea.io"},
+					sets.NewInt32(1),
+				},
+				{
+					"mockRule2",
+					[]string{"*antrea.io"},
+					sets.NewInt32(2),
+				},
+			},
+			map[string]dnsMeta{
+				"test.antrea.io": {},
+			},
+			"mockRule1",
+			[]string{"test.antrea.io"},
+			map[fqdnSelectorItem]sets.String{
+				selectorItem2: sets.NewString("mockRule2"),
+			},
+			map[string]map[fqdnSelectorItem]struct{}{
+				"test.antrea.io": {
+					selectorItem2: struct{}{},
+				},
+			},
+			true,
+		},
+		{
+			"test-4",
+			[]fqdnRuleAddArgs{
+				{
+					"mockRule1",
+					[]string{"maps.google.com"},
+					sets.NewInt32(1),
+				},
+				{
+					"mockRule2",
+					[]string{"*antrea.io$"},
```

Comment:
why is there a trailing dollar sign here? is it intentional? this is before conversion to regex

given how the test is written, it probably doesn't impact the result.